server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.search.SearchShardIterator;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.WeightedRoutingMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.rest.RestStatus;
+import org.opensearch.search.SearchShardTarget;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public enum FailAwareWeightedRouting {
+    INSTANCE;
+
+    private static final Logger logger = LogManager.getLogger(FailAwareWeightedRouting.class);
+
+    private final static List<RestStatus> internalErrorRestStatusList = List.of(
+        RestStatus.INTERNAL_SERVER_ERROR,
+        RestStatus.BAD_GATEWAY,
+        RestStatus.SERVICE_UNAVAILABLE,
+        RestStatus.GATEWAY_TIMEOUT
+    );
+
+    public static FailAwareWeightedRouting getInstance() {
+        return INSTANCE;
+
+    }
```

Comment:
I don't think this is the most apt way to create singletons, you could achieve it by a static property
```
    public static final FailAwareWeightedRouting INSTANCE = new FailAwareWeightedRouting();
```